src/storage/src/source/persist_source.rs

```diff
@@ -78,33 +77,39 @@ where
 
         let persist_client = persist_location.open().await?;
 
-        let (_write, read) =
-            persist_client.open::<Row, Row, Timestamp, Diff>(shard_id).await?;
+        let (_write, mut read) = persist_client
+            .open::<Row, Row, Timestamp, Diff>(shard_id)
+            .await?;
 
-            let mut snapshot_iter = read
-                .snapshot(as_of.clone())
-                .await
-                .expect("cannot serve requested as_of");
+        /// Aggressively downgrade `since`, to not hold back compaction.
+        read.downgrade_since(as_of.clone()).await;
 
+        let mut snapshot_iter = read
+            .snapshot(as_of.clone())
+            .await
+            .expect("cannot serve requested as_of");
 
-            // First, yield all the updates from the snapshot.
-            while let Some(next) = snapshot_iter.next().await {
-                yield ListenEvent::Updates(next);
-            }
+        // First, yield all the updates from the snapshot.
+        while let Some(next) = snapshot_iter.next().await {
+            yield ListenEvent::Updates(next);
+        }
 
-            // Then, listen continously and yield any new updates. This loop is expected to never
-            // finish.
-            let mut listen = read
-                .listen(as_of)
-                .await
-                .expect("cannot serve requested as_of");
-
-            loop {
-                let next = listen.next().await;
-                for event in next {
-                    yield event;
+        // Then, listen continously and yield any new updates. This loop is expected to never
+        // finish.
+        let mut listen = read
+            .listen(as_of)
+            .await
+            .expect("cannot serve requested as_of");
+
+        loop {
+            let next = listen.next().await;
+            for event in next {
+                if let ListenEvent::Progress(upper) = &event {
```

Comment:
@danhhz This got me thinking: all we're doing in the source is
1. create a listen
2. listen and downgrade as we get `Progress` events

It could be nice to have a method that consumes a `ReadHandle` to make a `Listen`, which would then only hold the since back as far as it needs to to make the listen work. This would free the user from having to worry about downgrading. Might be to spitball-y... 😅

Reply:
I've had the same thought myself! my hesitation in doing it so far has been that it only makes sense in a world where we have per-reader since capabilities. if we indeed switch to the shard-global downgrade-since model that storage has requested, then we'd just have to delete this code. OTOH, it's quite a nice ergonomics win in the meantime and probably not a ton of code. might be worth just doing anyway?

Reply:
Yeah, let me see once we finish with the current issues.

I think, though, that we can merge this PR once the other nits are resolved. And we add a `TODO` here.

Reply:
I've added a TODO in your name, hope that's fine :)

Reply:
> I've added a TODO in your name, hope that's fine :)

👌

Reply:
Ah, now that I thought about it some more: I _think_ we can downgrade the since all the way to the empty antichain after creating the listen. The listen will still work, and we're not blocking anything. (@danhhz if you could quickly double-check, that would be grand 🙏 )

Reply:
I've tried that and it looks like it works (the shard's `since` increases every loop iteration). I'll wait for @danhhz to tell us if this is something we can/should rely on.

Reply:
maybe I'm not quite understanding, but I don't think this is safe with the current implementation of listen. summary is that I think the following comment is incorrect

```
        // `listen` holds back the `since` frontier, so we can immediately downgrade the
        // one from `read` to the empty frontier.
```

Imagine if this was the last readhandle, then advancing the since to the empty antichain would allow compaction at any timestamp. the listen is required by the formalism to emit the data at the original timestamp, but compaction would be free to advance and consolidate data at will, so you'd have a race between the listen and compaction

if we do the idea aljoscha and I discussed above where the listen holds a capability on since and downgrades it, then we could do something like this, yeah

Reply:
I haven't thought about it deeply, but I think this is a correct way to get the shard-global since. The problem is that it doesn't prove anything wrt the issue I outlined :).

You could (and presumably do?) have another readhandle lying around somewhere that's holding a capability on the since and periodically downgrading it, which is why the since seems to be advancing (persist only does this in response to api usage, never e.g. internally on some schedule). If you didn't have that other handle, than downgrading the since here to the empty antichain could prevent the listener from working. And even though you do have it, if the listener falls behind whatever schedule the other readhandle is on, then once compaction is added, the listener may not be able to recover

Reply:
Ah, you are right! The storage controller keeps a read handle and periodically downgrades `since`. I was forgetting about that and attributed the downgrading to the `Listen` itself. Thanks for clearing this up!